File: Ahorn/entities/CustomDepthTileEntity.jl
module VivHelperDepthTileEntity
using ..Ahorn, Maple
using Ahorn.VivHelper

@mapdef Entity "VivHelper/CustomDepthTileEntity" DepthTileEntity(
x::Integer, y::Integer, width::Integer=8, height::Integer=8,
Depth::Integer=-10000)

const placements = Ahorn.PlacementDict(
    "Custom Depth Tile Entity (Viv's Helper)" => Ahorn.EntityPlacement(
        DepthTileEntity,
        "rectangle",
        Dict{String, Any}(),
        Ahorn.tileEntityFinalizer
    )
)

Ahorn.editingOptions(entity::DepthTileEntity) = Dict{String, Any}(
    "tiletype" => Ahorn.tiletypeEditingOptions(),
    "Depth" => VivHelper.Depths
)

Ahorn.minimumSize(entity::DepthTileEntity) = 8, 8
Ahorn.resizable(entity::DepthTileEntity) = true, true

Ahorn.selection(entity::DepthTileEntity) = Ahorn.getEntityRectangle(entity)

Ahorn.renderAbs(ctx::Ahorn.Cairo.CairoContext, entity::DepthTileEntity, room::Maple.Room) = Ahorn.drawTileEntity(ctx, room, entity)

end

File: Loenn/entities/CustomDepthTileEntity.lua
local vivUtil = require('mods').requireFromPlugin('libraries.vivUtil')
local fakeTilesHelper = require('helpers.fake_tiles')

local depthTileEntity = {}

depthTileEntity.name = "VivHelper/CustomDepthTileEntity"
depthTileEntity.depth = function(room,entity) return entity.Depth or -10000 end
depthTileEntity.placements = {
    name = "main",
    data = {
        tiletype = "3",
        Depth = -10000,
        width = 8,
        height = 8
    }
}


depthTileEntity.sprite = fakeTilesHelper.getEntitySpriteFunction("tiletype", false)
depthTileEntity.fieldInformation = function(entity)
    local orig = fakeTilesHelper.getFieldInformation("tiletype")(entity)
    orig["Depth"] = {fieldType = "integer"}
    return orig
end

return depthTileEntity